### PR TITLE
ck/#10-deserialiseModels

### LINKED ISSUE #10: De-serialize simbench models

### PR DESCRIPTION
**Pull request by ckittl**
Resolves #10 
De-serialising models from file

### CODE REVIEW

**Comment by ckittl** [comment]
!test

**Comment by ckittl** [comment]
!test

**Comment by ckittl** [comment]
!test

**Comment by ckittl** [comment]
!test

**Comment by ckittl** [comment]
!test

**Comment by sebastian-peter** [comment]
Reviewing this PR right now, will be done Wednesday (tomorrow) earliest.

File: src/main/scala/edu/ie3/simbench/model/datamodel/StudyCase.scala [hunk 1]
```diff
@@ -18,3 +24,40 @@ case class StudyCase(id: String,
                      pRes: BigDecimal,
                      vMSlack: BigDecimal)
     extends SimbenchModel
+
+case object StudyCase extends SimbenchCompanionObject[StudyCase] {
```
[the remark this answers is not in the document]
Comment (ckittl) [reply src/main/scala/edu/ie3/simbench/model/datamodel/StudyCase.scala]: Yes, some how. A case study examines different study cases. So the definition of multiple cases is mandatory do carry out an examination. :-)

File: src/main/scala/edu/ie3/simbench/config/SimbenchConfig.scala [hunk 2]
```diff
@@ -9,15 +9,38 @@ case class SimbenchConfig(
...
+    object Csv {
+      def apply(c: com.typesafe.config.Config): SimbenchConfig.Io.Csv = {
+        SimbenchConfig.Io.Csv(
+          fileEncoding =
+            if (c.hasPathOrNull("fileEncoding")) c.getString("fileEncoding")
+            else "UTF-8",
+          fileEnding =
+            if (c.hasPathOrNull("fileEnding")) c.getString("fileEnding")
+            else ".csv",
+          separator = c.getString("separator")
```
[the remark this answers is not in the document]
Comment (ckittl) [reply src/main/scala/edu/ie3/simbench/config/SimbenchConfig.scala]: Added it. Have initially not been sure, if the separator is consistent throughout all provided files.

File: src/main/scala/edu/ie3/simbench/io/CsvReader.scala [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    */
+  def read(desiredFields: Array[HeadLineField]): Vector[RawModelData] = {
+    val bufferedSource: BufferedSource =
+      io.Source.fromFile(filePath, fileEncoding)
+    try {
+      /* Build the file mapping */
+      val headLine = bufferedSource
+        .getLines()
+        .find(_ => true)
+        .getOrElse(throw IoException("The file does not contain any line."))
```
[the remark this answers is not in the document]
Comment (ckittl) [reply src/main/scala/edu/ie3/simbench/io/CsvReader.scala]: Not really. This line is meant to obtain the first line (head line) of the file and the easiest way without converting to another collection class. The `getOrElse` is only there to unbox the `Option`.

File: src/main/scala/edu/ie3/simbench/io/SimbenchReader.scala [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+      }
+
+    /* Creating study cases */
+    val studyCases = modelClassToRawData.get(classOf[StudyCase]) match {
+      case Some(rawDatas) => StudyCase.buildModels(rawDatas)
+      case None =>
+        logger.debug(
+          s"No information available for ${classOf[StudyCase].getSimpleName}")
+        Vector.empty
+    }
```
[the remark this answers is not in the document]
Comment (ckittl) [reply src/main/scala/edu/ie3/simbench/io/SimbenchReader.scala]: Great! I implemented it!

File: src/main/scala/edu/ie3/simbench/model/datamodel/Transformer3W.scala [hunk 5]
```diff
@@ -36,3 +39,21 @@ case class Transformer3W(id: String,
...
+  override def getFields: Array[HeadLineField] = Array.empty[HeadLineField]
+
+  /**
+    * Factory method to build one model from a mapping from field id to value
+    *
+    * @param rawData mapping from field id to value
+    * @return A model
+    */
+  override def buildModel(rawData: RawModelData): Transformer3W = ???
+}
```
[the remark this answers is not in the document]
Comment (ckittl) [reply src/main/scala/edu/ie3/simbench/model/datamodel/Transformer3W.scala]: Not really. The SimBench data set consists of a whole bunch of zip files. Anyways, I didn't found  any scheme, that I can be 100 % sure of, yet. I intend to resolve it as far as I get better information on that.

File: src/main/scala/edu/ie3/simbench/model/datamodel/types/Transformer3WType.scala [hunk 6]
```diff
@@ -76,3 +79,22 @@ case class Transformer3WType(id: String,
...
+    */
+  override def getFields: Array[HeadLineField] = Array.empty[HeadLineField]
+
+  /**
+    * Factory method to build one model from a mapping from field id to value
+    *
+    * @param rawData mapping from field id to value
+    * @return A model
+    */
+  override def buildModel(rawData: RawModelData): Transformer3WType = ???
```
[the remark this answers is not in the document]
Comment (ckittl) [reply src/main/scala/edu/ie3/simbench/model/datamodel/types/Transformer3WType.scala]: Same as explaned above...

**Comment by ckittl** [comment]
!test

File: src/main/scala/edu/ie3/simbench/model/datamodel/profiles/ResProfile.scala [hunk 7]
```diff
@@ -1,14 +1,156 @@
...
+  val WP3 = "WP3"
+  val WP4 = "WP4"
+  val WP5 = "WP5"
+  val WP6 = "WP6"
+  val WP7 = "WP7"
+  val WP8 = "WP8"
+  val WP9 = "WP9"
+  val WP10 = "WP10"
+  val WP11 = "WP11"
+  val WP12 = "WP12"
```
[the remark this answers is not in the document]
Comment (ckittl) [reply src/main/scala/edu/ie3/simbench/model/datamodel/profiles/ResProfile.scala]: I think an enumeration wouldn't make it much shorter. So for RES I would stick with this implementation.

File: src/main/scala/edu/ie3/simbench/io/SimbenchFileNamingStrategy.scala [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+      classOf[StudyCase] -> "StudyCases",
+      classOf[Transformer2W] -> "Transformer",
+      classOf[Transformer2WType] -> "TransformerType",
+      classOf[Transformer3W] -> "Transformer3W",
+      classOf[Transformer3WType] -> "Transformer3WType",
+      classOf[LoadProfile] -> "LoadProfile",
+      classOf[ResProfile] -> "RESProfile",
+      classOf[PowerPlantProfile] -> "PowerPlantProfile"
+    )
+  // TODO: Add naming strategy for NodePFResult
```
Comment (johanneshiry) [inline src/main/scala/edu/ie3/simbench/io/SimbenchFileNamingStrategy.scala]: pls address this issue or let me know if and why it needs to be still there

File: src/main/scala/edu/ie3/simbench/io/SimbenchFileNamingStrategy.scala [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+  /**
+    * Determine the file name based on the provided class
+    *
+    * @param clazz The class to determine the according file name for
+    * @return [[scala.util.Success]], if the file name can be determine, [[scala.util.Failure]], if not
+    */
+  def getFileName(clazz: Class[_]): Try[String] = {
```
Comment (johanneshiry) [inline src/main/scala/edu/ie3/simbench/io/SimbenchFileNamingStrategy.scala]: Why do you return a Try instead of an Option?

File: src/main/scala/edu/ie3/simbench/model/datamodel/Coordinate.scala [hunk 10]
```diff
@@ -15,3 +20,31 @@ case class Coordinate(id: String,
                       subnet: String,
                       voltLvl: Int)
     extends EntityModel
+
+case object Coordinate extends SimbenchCompanionObject[Coordinate] {
+  val X = "x"
```
Comment (johanneshiry) [inline src/main/scala/edu/ie3/simbench/model/datamodel/Coordinate.scala]:
```suggestion
  private val X = "x"
```

File: src/main/scala/edu/ie3/simbench/model/datamodel/Coordinate.scala [hunk 11]
```diff
@@ -15,3 +20,31 @@ case class Coordinate(id: String,
...
+  val Y = "y"
```
Comment (johanneshiry) [inline src/main/scala/edu/ie3/simbench/model/datamodel/Coordinate.scala]:
```suggestion
  private val Y = "y"
```

File: src/main/scala/edu/ie3/simbench/model/datamodel/PowerPlant.scala [hunk 12]
```diff
@@ -29,7 +34,7 @@ case class PowerPlant(id: String,
                       calculationType: CalculationType,
                       dspf: BigDecimal,
                       p: BigDecimal,
-                      q: BigDecimal,
+                      q: Option[BigDecimal],
```
Comment (johanneshiry) [inline src/main/scala/edu/ie3/simbench/model/datamodel/PowerPlant.scala]: Why did you change this to Option?

File: src/main/scala/edu/ie3/simbench/io/SimbenchFileNamingStrategy.scala (hunk 9 above)
Comment (ckittl) [reply src/main/scala/edu/ie3/simbench/io/SimbenchFileNamingStrategy.scala]: Because it is not really an option, if you have the file name or not. Or do you think of anything else, I did not see?

File: src/main/scala/edu/ie3/simbench/model/datamodel/PowerPlant.scala (hunk 12 above)
Comment (ckittl) [reply src/main/scala/edu/ie3/simbench/model/datamodel/PowerPlant.scala]: I didn't got it from specs, but in actual data, the field `qPP` is nullable. Then the reactive power is dependent of the calculation type, the power plant has.

File: src/main/scala/edu/ie3/simbench/model/datamodel/Coordinate.scala (hunk 10 above)
Comment (ckittl) [reply src/main/scala/edu/ie3/simbench/model/datamodel/Coordinate.scala]: Solved uniquely across all files.

File: src/main/scala/edu/ie3/simbench/model/datamodel/Coordinate.scala (hunk 11 above)
Comment (ckittl) [reply src/main/scala/edu/ie3/simbench/model/datamodel/Coordinate.scala]: Solved uniquely across all files.

**Comment by ckittl** [comment]
!test

**Comment by johanneshiry** [comment]
!test

**Comment by ckittl** [comment]
!test
